### PR TITLE
kill: implement a kill command

### PR DESCRIPTION
**Pull request by rminnich**
It takes as arguments -l, meaning list signals;
or -name, -s name, or --signal name and at least one pid.

TODO: write the test, which is a lot harder than it sounds at first.

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

File: cmds/kill/kill.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+)
+
+func usage() {
+	log.Fatalf("Usage: kill -l | kill -[s |-signal ]signame pid [pid...]")
+	os.Exit(1)
+}
+
+func main() {
+	a := os.Args
+	op := a[1]
```
Comment (hugelgupf) [inline cmds/kill/kill.go]: don't shorten os.Args -- go likes it short, but not always that short

```go
op := os.Args[1]
pids := os.Args[2:]
```

File: cmds/kill/kill.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+			usage()
+		}
+		fmt.Print(siglist())
+		os.Exit(1)
+	}
+
+	// N.B. Be careful if you want to change this. It has to continue to work if
+	// the signal is an arbitrary string. We take the path that the signal
+	// has to start with a -, or might be preceded by -s or --string.
+	if op[0] != '-' {
```
Comment (hugelgupf) [inline cmds/kill/kill.go]: Put this before op[0:2] above

File: cmds/kill/kill.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	s, ok := signums[op]
+	if !ok {
+		log.Fatalf("%v is not a valid signal", op)
+	}
+
+	if len(p) < 1 {
+		usage()
+	}
+	Kill(s, p...)
+
```
Comment (hugelgupf) [inline cmds/kill/kill.go]: extraneous empty line

File: cmds/kill/kill.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+)
+
+var (
+	l = flag.Bool("l", false, "list signal names")
```
Comment (hugelgupf) [inline cmds/kill/kill.go]: get rid of flag?

Comment (rminnich) [inline cmds/kill/kill.go]: done

File: cmds/kill/kill.go (hunk 1 above)
Comment (rminnich) [inline cmds/kill/kill.go]: done

File: cmds/kill/kill.go (hunk 2 above)
Comment (rminnich) [inline cmds/kill/kill.go]: It does not relate to that, I think it should be here as this is where we start processing signal names?

File: cmds/kill/kill.go (hunk 3 above)
Comment (rminnich) [inline cmds/kill/kill.go]: done

File: cmds/kill/kill.go (hunk 1 above)
Comment (rminnich) [inline cmds/kill/kill.go]: well, I don't like it as much now but ok.

**Comment by rminnich** [comment]
uh, oh, a big THANKS for the good review. We need more of them.

File: cmds/kill/kill.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+)
+
+func usage() {
+	log.Fatalf("Usage: kill -l | kill -[s |-signal ]signame pid [pid...]")
```
Comment (GanShun) [inline cmds/kill/kill.go]: default to sigterm?

File: cmds/kill/list_linux_amd64.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+		"SIGRTMAX-8":  syscall.Signal(56),
+		"SIGRTMAX-7":  syscall.Signal(57),
+		"SIGRTMAX-6":  syscall.Signal(58),
+		"SIGRTMAX-5":  syscall.Signal(59),
+		"SIGRTMAX-4":  syscall.Signal(60),
+		"SIGRTMAX-3":  syscall.Signal(61),
+		"SIGRTMAX-2":  syscall.Signal(62),
+		"SIGRTMAX-1":  syscall.Signal(63),
+		"SIGRTMAX":    syscall.Signal(64),
+		"1":           syscall.Signal(1),
```
Comment (GanShun) [inline cmds/kill/list_linux_amd64.go]: Is there not an easy way to autogenerate this? I guess it is easier to make changes in the future though.

File: cmds/kill/signaler_unix.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+func Kill(sig os.Signal, pids ...string) {
+	s := sig.(syscall.Signal)
+	for _, p := range pids {
+		pid, err := strconv.Atoi(p)
+		if err != nil {
+			log.Printf("Bad pid: %v", p)
+		}
+		i := syscall.Kill(pid, s)
+		if i != nil {
+			log.Printf("Kill %v with %v: %v", p, sig.String(), i)
```
Comment (GanShun) [inline cmds/kill/signaler_unix.go]: Change the print to properly signify that an error has occurred?

File: cmds/kill/list_linux_amd64.go (hunk 6 above)
Comment (rminnich) [inline cmds/kill/list_linux_amd64.go]: actually it was autogenerated from kill -l

but this stuff rarely changes, so I don't see a point in putting a lot of effort to auto generate it.

File: cmds/kill/signaler_unix.go (hunk 7 above)
Comment (rminnich) [inline cmds/kill/signaler_unix.go]: I got rid of the print entirely, it did not belong in this function.

File: cmds/kill/list_linux_amd64.go (hunk 6 above)
Comment (GanShun) [inline cmds/kill/list_linux_amd64.go]: Fair enough.

**Comment by rminnich** [comment]
so, you good on this, or ...

**Comment by GanShun** [comment]
Did you make the change for the default? if you don't want to do it now that's fine to. maybe put in a TODO?

**Comment by rminnich** [comment]
I wonder if I forgot to push?

File: cmds/kill/kill.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+package main
+
+import (
+	"fmt"
+	"os"
+)
+
+func usage() {
+	die("Usage: kill -l | kill -[s |-signal ]signame pid [pid...]")
```
Comment (GanShun) [inline cmds/kill/kill.go]: change this print then. otherwise I'm good with this

**Pull request merged**